docs/rfcs/20220131_raft_snapshots.md

```diff
@@ -0,0 +1,149 @@
+
+#### Modify the `append_entries_request`
+
+We may modify the append_enties_request to include an offset of the last compaction. When a follower sees that last compaction less than current offset it knows that the latest batches don't have holes and it's safe to serve the requests.
+
+When we modify `append_entries_request` or introduce and switch to the new API we need to thing of the update story. In order to avoid shutting down a cluster for the update we need to put the change behind a feature flag, update the brokers one by one and only then to active the feature.
+
+## Approach #2: Ship the snapshot (cached state) via the new API
+
+An alternative idea is to use a side channel instead of the main replication path to ship the snapshot. With this approach we don't shift the offset and it eventually opens a way to get rid of the offset translation layer.
+
+### Overview
+
+When a leader decides to compact the log at offset X it:
+
+  1. makes a snapshot and sends it to a majority of nodes via new API
+  2. waits until it's safe to do compaction:
+    a. segment containing X is closed
+    b. tx front is passed the end of the segment
+  3. starts include the new offset X to the `append_entries_request`
+  4. performs the compaction
+ 
+When a follower observes a compacted offset past its snapshot via `append_entries_request` it:
+
+  1. queries all nodes and finds a node which has the snapshot (there is at least a majority of such nodes so availability stays the same as with Raft)
+  2. downloads the snapshot
+  3. applies the snapshot
+  4. rewinds the snapshot to the snapshot's offset
+  5. acknowledges the `append_entries_request`
+
+It's safe to serve read from a replica bounded by the last snapshot's offset or by current offset if the compacted offset is equal to the last snapshot.
+
+## Interaction with other features
+
+The RFC doesn't cover SI code path but it seems that it should compatible with the changes, please comment if you see any concerns.
+
+## Telemetry & Observability
+
+We may monitor the size of the snapshot and the lag between the offset a follower saw and the consumable offset.
+
+## Rationale and Alternatives
+
+Shipping via the log gives us results faster and shipping via a new API takes more time to implement. However if we don't want to limit the horizon of read replicas by the last compaction time (last snapshot) we need to modify `append_entries_request` to include compacted offset and it is inline with shipping snapshots via the new API. So we may start with the fast solution and then morph it into the one which doesn't shift the offsets without reworking everything.
+
+  1. ship snapshot via the log
+  2. modify `append_entries_request`
+  3. ship log via the new API
+
+With this plan we'll get actionable results at each step. The alternative to jump directly to step three.
+
+## Unresolved questions
+
+Should we do anything with respect to
+- integration with SI
+- integration with WASM transform
```

Comment:
I think all WASM would have to do is set it's maximum read offset to the newly proposed offset (min compacted offset).